workbench/src/wb_project.c

```diff
@@ -0,0 +1,1491 @@
+ * project structure @a prj.
+ *
+ * @param prj      The project
+ * @param filename File to load
+ * @param error    Location to store error info at
+ * @return TRUE on success, FALSE otherwise
+ *
+ **/
+gboolean wb_project_load(WB_PROJECT *prj, gchar *filename, GError **error)
+{
+	gboolean success = FALSE;
+
+	if (prj != NULL)
+	{
+		GKeyFile *kf;
+		guint	 index;
+		gchar    *contents;
+		gchar    key[100];
+		gsize    length;
+
+		if (!g_file_get_contents (filename, &contents, &length, error))
+		{
+			return FALSE;
+		}
+
+		kf = g_key_file_new ();
+
+		if (!g_key_file_load_from_data (kf, contents, length,
+					G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS,
+					error))
+		{
+			g_key_file_free (kf);
+			g_free (contents);
+			return FALSE;
+		}
+
+		if (g_key_file_has_group (kf, "Workbench"))
+		{
+			WB_PROJECT_DIR *new_dir;
+			gchar *str;
+			gchar **splitv, **bookmarks_strings;
+
+			/* Load project bookmarks from string list */
+			bookmarks_strings = g_key_file_get_string_list (kf, "Workbench", "Bookmarks", NULL, error);
+			if (bookmarks_strings != NULL)
+			{
+				gchar **file, *abs_path;
+
+				file = bookmarks_strings;
+				while (*file != NULL)
+				{
+					abs_path = get_combined_path(prj->filename, *file);
+					if (abs_path != NULL)
+					{
+						wb_project_add_bookmark_int(prj, abs_path);
+						g_free(abs_path);
+					}
+					file++;
+				}
+				g_strfreev(bookmarks_strings);
+			}
+
+			/* Load project dirs */
+			for (index = 1 ; index < 1025 ; index++)
+			{
+				g_snprintf(key, sizeof(key), "Dir%u-BaseDir", index);
+				if (!g_key_file_has_key (kf, "Workbench", key, NULL))
+				{
+					break;
+				}
+
+				str = g_key_file_get_string(kf, "Workbench", key, NULL);
+				if (str == NULL)
+				{
+					break;
+				}
+				new_dir = wb_project_add_directory_int(prj, str, FALSE);
+				if (new_dir == NULL)
+				{
+					break;
+				}
+
+				g_snprintf(key, sizeof(key), "Dir%u-FilePatterns", index);
+				str = g_key_file_get_string(kf, "Workbench", key, NULL);
+				if (str != NULL)
+				{
+					splitv = g_strsplit (str, ";", -1);
+					wb_project_dir_set_file_patterns(new_dir, splitv);
+				}
+				g_free(str);
+
+				g_snprintf(key, sizeof(key), "Dir%u-IgnoredDirsPatterns", index);
+				str = g_key_file_get_string(kf, "Workbench", key, NULL);
+				if (str != NULL)
+				{
+					splitv = g_strsplit (str, ";", -1);
+					wb_project_dir_set_ignored_dirs_patterns(new_dir, splitv);
+				}
+				g_free(str);
+
+				g_snprintf(key, sizeof(key), "Dir%u-IgnoredFilePatterns", index);
+				str = g_key_file_get_string(kf, "Workbench", key, NULL);
+				if (str != NULL)
+				{
+					splitv = g_strsplit (str, ";", -1);
+					wb_project_dir_set_ignored_file_patterns(new_dir, splitv);
+				}
+				g_free(str);
+			}
+		}
+
+		g_key_file_free(kf);
+		g_free (contents);
+		success = TRUE;
+	}
+	else if (error != NULL)
+	{
+		g_set_error (error, 0, 0,
+					 "Internal error: param missing (file: %s, line %d)",
+					 __FILE__, __LINE__);
```

Comment:
Here again, it probably rather calls for a `g_return_val_if_fail(prf != NULL, FALSE)` at the start of the function.  `GErrors` are supposed to carry runtime errors (like IO error, file missing, etc.), not programming errors (invalid parameters, etc.).